### PR TITLE
feat(optimizer): support index merge optimization

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

### What's Index Merge Optimization?
https://dev.mysql.com/doc/refman/8.0/en/index-merge-optimization.html
https://www.postgresql.org/docs/current/indexes-bitmap-scans.html

- Support index merge optimization for batch query.
- To deal with query condition like `a = 1 or c = 1`, we need index merge, because only accessing one index can't provide an efficient execution plan. At least, we need to access two indexes, one for condition `a = 1` and another for `c = 1`.
- The basic algorithm of index merge optimization is to find out the index we need to merge and then fetch the `row id of the primary table` from index. Finally lookup the primary table by row id fetched before.

### Example
```SQL
create table t1 (a int, b numeric, c bigint);
create index idx1 on t1(a);
create index idx2 on t1(b, a);
create index idx3 on t1(c);
explain select * from t1 where a = 1 or c = 1

BatchLookupJoin { type: Inner, predicate: idx1.t1._row_id = t1._row_id AND ((t1.a = 1:Int32) OR (t1.c = 1:Int32)), output: [t1.a, t1.b, t1.c] }
  BatchHashAgg { group_key: [idx1.t1._row_id], aggs: [] }
    BatchUnion { all: true }
      BatchExchange { order: [], dist: Single }
        BatchScan { table: idx1, columns: [idx1.t1._row_id], scan_ranges: [idx1.a = Int32(1)], distribution: SomeShard }
      BatchExchange { order: [], dist: Single }
        BatchScan { table: idx3, columns: [idx3.t1._row_id], scan_ranges: [idx3.c = Int64(1)], distribution: SomeShard }
```

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
#4339

### CODE REVIEW

File: src/frontend/src/optimizer/rule/index_selection.rs [hunk 1]
```diff
@@ -81,8 +90,7 @@ impl Rule for IndexSelectionRule {
             return None;
         }
 
-        let mut primary_table_scan_io_estimator = TableScanIoEstimator::new(logical_scan);
-        let primary_cost = primary_table_scan_io_estimator.estimate(logical_scan.predicate());
+        let primary_cost = self.estimate_table_scan_cost(logical_scan);
```
Comment (BowenXiao1999) [inline src/frontend/src/optimizer/rule/index_selection.rs:93]: actually I previously think the cost of scanning from primary should determined by index cost 🤔 .

File: src/frontend/src/optimizer/rule/index_selection.rs [hunk 2]
```diff
@@ -237,6 +237,359 @@ impl IndexSelectionRule {
...
+
+        Some((
+            lookup_join,
+            index_access_cost.mul(&IndexCost::new(LOOKUP_COST_CONST)),
+        ))
+    }
+
+    /// Generate possible paths that can be used to access.
+    /// The schema of output is the order key of primary table, so it can be used to lookup primary
+    /// table later.
```
Comment (BowenXiao1999) [inline src/frontend/src/optimizer/rule/index_selection.rs:308]: Difference and connection with gen_index_path?

File: src/frontend/src/optimizer/rule/index_selection.rs [hunk 3]
```diff
@@ -237,6 +237,359 @@ impl IndexSelectionRule {
...
+            }
+        }
+
+        result
+    }
+
+    /// Clustering the disjunction by column index
+    /// a = 1 or b = 2 or b = 3 -> <a, (a = 1)>, <b, (b = 2 or b = 3)>
+    /// a = 1 or (b = 2 and c = 3) -> <a, (a = 1)>, <None, (b = 2 and c = 3)>
+    fn clustering_disjunction(
```
Comment (BowenXiao1999) [inline src/frontend/src/optimizer/rule/index_selection.rs:350]: Actually do not get the meaning. Why set b -> None if there is a c?

**Review by BowenXiao1999 — COMMENTED** [review]
Now we use a Union + HashAgg to do the dedup? Are there alternatives plan?

File: src/frontend/src/optimizer/rule/index_selection.rs (hunk 1 above)
Comment (chenzl25) [reply src/frontend/src/optimizer/rule/index_selection.rs:93]: Primary table is also an index.

**Comment by chenzl25** [comment]
> Now we use a Union + HashAgg to do the dedup? Are there alternatives plan?

If we don't invent some other operators, Agg and Union are the only choice.

**Review by BowenXiao1999 — APPROVED** [review]

**Pull request merged**